api/v1alpha1/nutanix_clusterconfig_types.go

```diff
@@ -76,3 +84,39 @@ func (s NutanixPrismCentralEndpointSpec) ParseURL() (string, int32, error) {
 
 	return hostname, int32(port), nil
 }
+
+// NutanixFailureDomains is a list of FDs.
+type NutanixFailureDomains []NutanixFailureDomain
+
+// NutanixFailureDomain configures failure domain information for Nutanix.
+type NutanixFailureDomain struct {
+	// name defines the unique name of a failure domain.
+	// Name is required and must be at most 64 characters in length.
+	// It must consist of only lower case alphanumeric characters and hyphens (-).
+	// It must start and end with an alphanumeric character.
+	// This value is arbitrary and is used to identify the failure domain within the platform.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=64
+	// +kubebuilder:validation:Pattern=`[a-z0-9]([-a-z0-9]*[a-z0-9])?`
+	Name string `json:"name"`
+
+	// cluster is to identify the cluster (the Prism Element under management of the Prism Central),
+	// in which the Machine's VM will be created. The cluster identifier (uuid or name) can be obtained
+	// from the Prism Central console or using the prism_central API.
+	// +kubebuilder:validation:Required
+	Cluster NutanixResourceIdentifier `json:"cluster"`
+
+	// subnets holds a list of identifiers (one or more) of the cluster's network subnets
+	// for the Machine's VM to connect to. The subnet identifiers (uuid or name) can be
+	// obtained from the Prism Central console or using the prism_central API.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	// +listType=map
+	// +listMapKey=type
+	Subnets []NutanixResourceIdentifier `json:"subnets"`
+
+	// indicates if a failure domain is suited for control plane nodes
+	// +kubebuilder:validation:Required
+	ControlPlane bool `json:"controlPlane,omitempty"`
```

Comment:
Is it valid to have all FD with this value set to `false`?

Reply:
Yes, it is valid in backend but we can have UI/CLI make it mandatory to provide one fd with this to be true. Ideally it will create controlPlane on provided cluster and subnet in the respective machine template if no fd has ControlPlane is true. if nmt does not have cluster and subnet provided then it will fail with this error https://github.com/nutanix-cloud-native/cluster-api-provider-nutanix/blob/v1.4.0-alpha.2/controllers/nutanixmachine_controller.go#L836.